fedimintd/src/ui.rs

```diff
@@ -52,26 +52,42 @@ async fn home_page(axum::extract::State(_): axum::extract::State<MutableState>)
 #[derive(Template)]
 #[template(path = "run.html")]
 struct RunTemplate {
-    connection_string: String,
-    has_connection_string: bool,
+    state: RunTemplateState,
+}
+
+enum RunTemplateState {
+    DkgNotStarted,
+    DkgInProgress,
+    DkgDone(String),   // connnection string
+    DkgFailed(String), // error
+    LocalIoError(String),
 }
 
 async fn run_page(axum::extract::State(state): axum::extract::State<MutableState>) -> RunTemplate {
     let state = state.lock().await;
-    let path = state.data_dir.join("client.json");
-    let connection_string: String = match std::fs::File::open(path) {
-        Ok(file) => {
-            let cfg: ClientConfig =
-                serde_json::from_reader(file).expect("Could not parse cfg file.");
-            let connect_info = WsFederationConnect::from(&cfg);
-            serde_json::to_string(&connect_info).expect("should deserialize")
-        }
-        Err(_) => "".into(),
-    };
 
     RunTemplate {
-        connection_string: connection_string.clone(),
-        has_connection_string: !connection_string.is_empty(),
+        state: match state.dkg_state {
+            Some(DkgState::Success) => {
+                let path = state.data_dir.join("client.json");
+                match std::fs::File::open(path) {
+                    Ok(file) => match serde_json::from_reader(file) {
+                        Ok(cfg) => {
+                            let connect_info = WsFederationConnect::from(&cfg);
+
+                            RunTemplateState::DkgDone(
+                                serde_json::to_string(&connect_info).expect("should deserialize"),
+                            )
+                        }
+                        Err(e) => RunTemplateState::LocalIoError(e.to_string()),
+                    },
+                    Err(e) => RunTemplateState::LocalIoError(e.to_string()),
+                }
```

Comment:
Could this be an extra fn with early exit to avoid the nesting?

Reply:
Right, but I'll ignore as I don't have time to retest it right now. Added a TODO comment to refactor.